### PR TITLE
Adding Modal component

### LINKED ISSUE #15432: [Ext] Create component: Modal
### Description
Create a reusable UI component for `Modal` according to the insight report

Tasks
- [x] Merge https://github.com/MetaMask/metamask-extension/pull/18311
- [x] Merge https://github.com/MetaMask/metamask-extension/pull/18161
- [x] Merge https://github.com/MetaMask/metamask-extension/pull/18175
- [x] https://github.com/MetaMask/metamask-extension/pull/18979

### References
[Modal Audit FigJam](https://www.figma.com/file/hxYqloYgmVcgsoiVqmGZ8K/Modal-Audit?node-id=0%3A1&t=WcWEbEnjEkGuOdXr-1)
[Figma component](https://www.figma.com/file/HKpPKij9V3TpsyMV1TpV7C/DS-Components?node-id=3180%3A13642)
[UI component guidelines](https://www.notion.so/UI-Components-5ccd2bf83eb8441892a0b72c0d8929e1)
[Styling guidelines](https://www.notion.so/Extension-Frontend-Engineering-Guide-b82ddb3e14004b5db3799a9b446294a9#49aa2ef7b7aa40178ff918043df71c34)
Testing guidelines: TBC

### Files needed
- `index.js`
- `index.scss`
- `modal.js`
- `modal.stories.js`
- `modal.test.js`
- `README.mdx`

### Technical details
TBC

### Acceptance criteria
- [ ] Uses color, typography, shadows design tokens 
- [ ] Uses semantic html
- [ ] PropTypes have descriptions
- [ ] Has storybook story with controls
- [ ] Has documentation in MDX
- [ ] Has unit tests and 90% coverage
- [ ] Works in Chrome and Firefox
- [ ] Performance tested: no unnecessary re-renders or other performance concerns

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## Explanation
Adds new `Modal` component that uses context to provide state to the sub components from the parent component. This also updates the `ModalContent` component so options are only set once and then handled through the `Modal` context

* Fixes #15432

Dependencies
- https://github.com/MetaMask/metamask-extension/pull/19034
- https://github.com/MetaMask/metamask-extension/pull/18979

### Before
Current `ModalContent`

https://github.com/MetaMask/metamask-extension/assets/8112138/9bcd9354-b67e-403a-9e32-1c78a78365d5

### After
Updated `ModalContent`

https://github.com/MetaMask/metamask-extension/assets/8112138/f7ae7423-628d-4a2e-9dc3-6534437b3f86

New `Modal` component and documentation

https://github.com/MetaMask/metamask-extension/assets/8112138/9615b660-81f9-4292-a29c-60638dea42e9

## Manual Testing Steps

<!--
How should reviewers and QA manually test your changes? For instance:

- Go to the storybook build in this PR
- Search `Modal` component and go to one in the Component Library
- Check controls, stories and documentation

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/component-library/button/button.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ import { ButtonLink } from '../button-link';
 
 import { BUTTON_VARIANT } from './button.constants';
 
-export const Button = ({ variant, ...props }) => {
+export const Button = ({ variant = BUTTON_VARIANT.PRIMARY, ...props }) => {
```
Comment (georgewrmarshall) [inline ui/components/component-library/button/button.js:10]: Move to new PR

File: ui/components/component-library/modal-content/README.mdx [hunk 2]
```diff
@@ -18,7 +18,7 @@ The `ModalContent` accepts all props below as well as all [Box](/docs/components
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-content/README.mdx:18]: Update description of `ModalContent` to:

```
`ModalContent` is the container for the modal dialog's content. It uses context supplied by the `Modal` component and cannot be used without it.

File: ui/components/component-library/modal-content/__snapshots__/modal-content.test.tsx.snap [hunk 3]
```diff
@@ -1,11 +1,3 @@
...
-exports[`ModalContent should match snapshot 1`] = `
-<div>
-  <div
-    class="box mm-modal-content mm-modal-content--size-sm box--padding-4 box--flex-direction-row box--width-full box--background-color-background-default box--rounded-lg"
-  >
-    test
-  </div>
-</div>
-`;
+exports[`ModalContent should match snapshot 1`] = `<div />`;
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-content/__snapshots__/modal-content.test.tsx.snap:3]: This seems funky. Check test

File: ui/components/component-library/modal-header/__snapshots__/modal-header.test.tsx.snap [hunk 4]
```diff
@@ -9,11 +9,11 @@ exports[`ModalHeader should render ModalHeader correctly 1`] = `
...
-      <h4
+      <header
         class="box mm-text mm-text--heading-sm mm-text--text-align-center box--flex-direction-row box--color-text-default"
       >
         Modal header
-      </h4>
+      </header>
     </div>
   </div>
 </div>
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/__snapshots__/modal-header.test.tsx.snap:19]: Move to separate PR

File: ui/components/component-library/modal-header/modal-header.tsx [hunk 5]
```diff
@@ -50,7 +57,11 @@ export const ModalHeader: React.FC<ModalHeaderProps> = ({
...
       {typeof children === 'string' ? (
-        <Text variant={TextVariant.headingSm} textAlign={TextAlign.Center}>
+        <Text
+          as={ValidTag.Header}
+          variant={TextVariant.headingSm}
+          textAlign={TextAlign.Center}
+        >
           {children}
         </Text>
       ) : (
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/modal-header.tsx:67]: Move to separate PR

File: ui/components/component-library/modal-overlay/__snapshots__/modal-overlay.test.tsx.snap [hunk 6]
```diff
@@ -3,6 +3,7 @@
 exports[`ModalOverlay should match snapshot 1`] = `
 <div>
   <div
+    aria-hidden="true"
     class="box mm-modal-overlay box--flex-direction-row box--width-full box--height-full box--background-color-overlay-default"
   />
 </div>
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-overlay/__snapshots__/modal-overlay.test.tsx.snap:9]: Move to separate PR

File: ui/components/component-library/modal-overlay/modal-overlay.scss [hunk 7]
```diff
@@ -1,7 +1,8 @@
 .mm-modal-overlay {
-  position: absolute;
+  position: fixed;
   top: 0;
   left: 0;
   right: 0;
   bottom: 0;
+  z-index: $modal-z-index;
 }
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-overlay/modal-overlay.scss:8]: Move to separate PR

File: ui/components/component-library/modal-overlay/modal-overlay.tsx [hunk 8]
```diff
@@ -21,6 +21,7 @@ export const ModalOverlay: React.FC<ModalOverlayProps> = ({
     width={BLOCK_SIZES.FULL}
     height={BLOCK_SIZES.FULL}
     onClick={onClick}
+    aria-hidden="true"
     {...props}
   />
 );
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-overlay/modal-overlay.tsx:27]: Move to separate PR

File: ui/components/component-library/modal/modal.test.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        isOpen
+        className="test-class"
+        data-testid="modal"
+      >
+        <div>modal dialog</div>
+      </Modal>,
+    );
+    expect(getByTestId('modal')).toHaveClass('mm-modal test-class');
+  });
+});
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/modal.test.tsx:46]: Add some tests to check context is being passed down for all props

File: ui/components/component-library/modal/modal.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import classnames from 'classnames';
+
+import { ModalProps } from './modal.types';
+import { ModalContext } from './modal.context';
+
+export const Modal: React.FC<ModalProps> = ({
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/modal.tsx:8]: Forward ref to div. Should be as standard for all of our components probably

File: ui/components/component-library/modal/modal.types.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+   * @default true
+   */
+  isClosedOnOutsideClick?: boolean;
+  /**
+   * closeOnEscape enables the ability to close the modal when the user presses the escape key
+   * If this is disabled there should be a close button in the modal or allow keyboard only users to close the modal with a button that is accessible via the tab key
+   *
+   * @default true
+   */
+  isClosedOnEscapeKey?: boolean;
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/modal.types.ts:32]: There is nothing specific to the modal about these props but maybe that's ok?

File: ui/components/component-library/text/text.types.ts [hunk 12]
```diff
@@ -41,6 +41,7 @@ export enum ValidTag {
   Ul = 'ul',
   Label = 'label',
   Input = 'input',
+  Header = 'header',
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.types.ts:44]: Move to separate PR

File: ui/components/ui/popover/index.scss [hunk 13]
```diff
@@ -51,7 +51,7 @@
     display: flex;
     justify-content: center;
     align-items: center;
-    z-index: 1050;
+    z-index: $modal-z-index;
```
Comment (georgewrmarshall) [inline ui/components/ui/popover/index.scss:54]: Move to separate PR

File: ui/css/design-system/z-index.scss [hunk 14]
```diff
@@ -10,3 +10,4 @@ $main-container-z-index: 18;
 $send-card-z-index: 20;
 $sidebar-z-index: 26;
 $sidebar-overlay-z-index: 25;
+$modal-z-index: 1050;
```
Comment (georgewrmarshall) [inline ui/css/design-system/z-index.scss:13]: Move to separate PR

File: ui/components/component-library/modal-focus/README.mdx [hunk 15]
```diff
@@ -0,0 +1,248 @@
+import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-focus/README.mdx:1]: Will be removed once https://github.com/MetaMask/metamask-extension/pull/18979 is merged

File: ui/components/component-library/modal-focus/index.ts [hunk 16]
```diff
@@ -0,0 +1,2 @@
+export { ModalFocus } from './modal-focus';
+export type { ModalFocusProps } from './modal-focus.types';
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-focus/index.ts:2]: Will be removed once https://github.com/MetaMask/metamask-extension/pull/18979 is merged

File: ui/components/component-library/modal-focus/modal-focus.stories.tsx [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+      )}
+    </>
+  );
+};
+
+export const AutoFocus = Template.bind({});
+AutoFocus.args = {
+  autoFocus: false,
+  children: <p>auto focus set to false</p>,
+};
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-focus/modal-focus.stories.tsx:158]: Will be removed once https://github.com/MetaMask/metamask-extension/pull/18979 is merged

File: ui/components/component-library/modal-focus/modal-focus.tsx [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+      onDeactivation={onDeactivation}
+      returnFocus={returnFocus}
+      {...props}
+    >
+      {children}
+    </FocusTrap>
+  );
+};
+
+ModalFocus.displayName = 'ModalFocus';
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-focus/modal-focus.tsx:46]: Will be removed once https://github.com/MetaMask/metamask-extension/pull/18979 is merged

File: ui/components/component-library/modal-focus/modal-focus.types.ts [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+   */
+  children: React.ReactNode;
+  /**
+   * If `true`, the first focusable element within the `children`
+   * will auto-focused once `ModalFocus` mounts
+   *
+   * @default false
+   */
+  autoFocus?: boolean;
+}
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-focus/modal-focus.types.ts:35]: Will be removed once https://github.com/MetaMask/metamask-extension/pull/18979 is merged

File: ui/components/component-library/modal-header/__snapshots__/modal-header.test.tsx.snap (hunk 4 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal-header/__snapshots__/modal-header.test.tsx.snap:19]: Done https://github.com/MetaMask/metamask-extension/pull/19034

File: ui/components/component-library/modal-header/modal-header.tsx (hunk 5 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal-header/modal-header.tsx:67]: Done https://github.com/MetaMask/metamask-extension/pull/19034

File: ui/components/component-library/modal-overlay/__snapshots__/modal-overlay.test.tsx.snap (hunk 6 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal-overlay/__snapshots__/modal-overlay.test.tsx.snap:9]: Done https://github.com/MetaMask/metamask-extension/pull/19034

File: ui/components/component-library/modal-overlay/modal-overlay.scss (hunk 7 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal-overlay/modal-overlay.scss:8]: Done https://github.com/MetaMask/metamask-extension/pull/19034

File: ui/components/component-library/modal-overlay/modal-overlay.tsx (hunk 8 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal-overlay/modal-overlay.tsx:27]: Done https://github.com/MetaMask/metamask-extension/pull/19034

File: ui/components/component-library/text/text.types.ts (hunk 12 above)
Comment (georgewrmarshall) [reply ui/components/component-library/text/text.types.ts:44]: Done https://github.com/MetaMask/metamask-extension/pull/19034

File: ui/components/ui/popover/index.scss (hunk 13 above)
Comment (georgewrmarshall) [reply ui/components/ui/popover/index.scss:54]: Done https://github.com/MetaMask/metamask-extension/pull/19034

File: ui/css/design-system/z-index.scss (hunk 14 above)
Comment (georgewrmarshall) [reply ui/css/design-system/z-index.scss:13]: Done https://github.com/MetaMask/metamask-extension/pull/19034

File: ui/components/component-library/modal-content/__snapshots__/modal-content.test.tsx.snap (hunk 3 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal-content/__snapshots__/modal-content.test.tsx.snap:3]: Turns out this would be the expected result for the container snapshot. Have updated it to use getByTestId so it renders the wrapping div and className

File: ui/components/component-library/modal-content/README.mdx (hunk 2 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal-content/README.mdx:18]: Updated

File: ui/components/component-library/modal/modal.tsx (hunk 10 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal/modal.tsx:8]: Updated

**Review by darkwing — COMMENTED** [review]
Overall looks good but shouldn't `ModalContent` -> `Children` stay within the height dimension of the window, with scrolling inner content?  Or am I misunderstanding `Modal`?

**Comment by georgewrmarshall** [comment]
> Overall looks good but shouldn't `ModalContent` -> `Children` stay within the height dimension of the window, with scrolling inner content? Or am I misunderstanding `Modal`?

That is a great point @darkwing I based it on the Chakra UI modal which behaves the same an IMO operates more like a page and allows more content to be visible but I will check with design and see which behaviour they prefer.

**Comment by darkwing** [comment]
> I think the scrolling could be on a case by case basis? WDYT?

Good point.  Should we implement that prop now while we can, or wait until we _need_ it?  I think having it as a single prop, much like `ellipsis`, would be best, instead of a combo of them for each case.

**Comment by georgewrmarshall** [comment]
I'm hesitant to add it to this PR because of it's size, I think there is a current work around to be able to use it immediately and I've created an issue and added it to the next sprint to work on. Would it be ok to get this in and add it as another feature of the Modal?

https://github.com/MetaMask/metamask-extension/issues/19190

File: ui/components/component-library/modal-content/modal-content.tsx [hunk 20]
```diff
@@ -1,37 +1,106 @@
...
+      if (
+        isClosedOnOutsideClick &&
+        modalDialogRef?.current &&
+        !modalDialogRef.current.contains(event.target as Node)
+      ) {
+        onClose();
+      }
+    };
+
+    useEffect(() => {
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/modal-content/modal-content.tsx:56]: Hmm I don't think we would want to put the useEffect inside a conditional statement? useEffect is meant to handle side effects on every render or in response to specific dependencies. By putting it in a conditional it could lead to inconsistent behaviour depending on the condition and how it changes over renders. I could put the conditional inside of the useEffect and add `isClosedOnEscapeKey` and `isClosedOnOutsideClick` to the dependency array.  e.g.

```
  if (isClosedOnEscapeKey) {
    document.addEventListener('keydown', handleEscKey);
  }

  if (isClosedOnOutsideClick) {
    document.addEventListener('mousedown', handleClickOutside);
  }

  return () => {
    if (isClosedOnEscapeKey) {
      document.removeEventListener('keydown', handleEscKey);
    }

    if (isClosedOnOutsideClick) {
      document.removeEventListener('mousedown', handleClickOutside);
    }
  };
```

But I think the performance improvements would be pretty negligible? These event listeners would only be added and removed every time a Modal is open. But it would be good to monitor

**Review by garrettbear — APPROVED** [review]

**Review by NidhiKJha — APPROVED** [review]
